### PR TITLE
Merge standalone Search Profile v1p1 extension spec document to develop branch

### LINKED ISSUE #308: Create standalone Search Profile
Per https://github.com/IMSGlobal/caliper-central/issues/128

Draft a Search Profile that defines the following new terms:
1. A new `SearchEvent`.
2. A `Query` entity (subTypeOf Entity)
3. A `searchString` property.  Domain: `Query`; range: `xsd:string`

Also create a new Search profile scoped Caliper context.

### PR DESCRIPTION
**Pull request by arwhyte**
Resolves #308.

### CODE REVIEW

**Comment by arwhyte** [comment]
Raw git view: https://rawgit.com/IMSGlobal/caliper-spec/c791e81d8099df7a6f367cf448b59c038f545428/profiles/caliper-profile-search-v1p1.html

**Comment by arwhyte** [comment]
*TODO*: I think we should add a section that highlights use of a profile-scoped JSON-LD context.  Relying solely on the examples I suspect will prove an insufficient hint.  So add something like:

**SearchEvent JSON-LD context**
IMS Global provides a remote Search Profile-scoped JSON-LD context document (http://purl.imsglobal.org/ctx/caliper/v1p1/SearchProfile-extension) for mapping `SearchEvent1` terms to IRIs.  Caliper message producers MUST reference this context in every Caliper 1.1 `SearchEvent` and/or `Query` entity *describe* created and serialized for transmission to a target endpoint.

** Example **
```
{
  "@context": "http://purl.imsglobal.org/ctx/caliper/v1p1/SearchProfile-extension",
 . . .
}

I've scoped the last sentence to Caliper 1.1 deliberately; once Caliper 1.2 is released event producers will be able to issue `SearchEvent` and `Query` describes referencing a 1.2 context.

**Comment by arwhyte** [comment]
@johnpj asked if it would be possible to provide a list of search results.  

Below is an example of a SearchEvent that includes a “generated” DigitalResourceCollection of search results.  The example makes use of existing vocabulary; we could consider creating a subtype of DigitalResourceCollection, say, SearchResults, to more clearly mark the generated value as a set of search results.  Ideally, the SearchResults entity would express additional properties relevant to a collection of search results.

```
{
  "@context": "http://purl.imsglobal.org/ctx/caliper/v1p1/SearchProfile-extension",
  "id": "urn:uuid:cb3878ed-8240-4c6d-9fee-77221810f5e4",
  "type": "SearchEvent",
  "actor": {
    "id": "https://example.edu/users/554433",
    "type": "Person"
  },
  "action": "Searched",
  "object": {
    "id": "https://example.edu/catalog",
    "type": "SofwareApplication"
  },
  "eventTime": "2018-11-15T10:05:00.000Z”,

  . . .


  "generated": {
    "id": "https://example.edu/catalog/search/334455?query=IMS%20AND%20%28Caliper%20OR%20Analytics%29",
    "type": "DigitalResourceCollection",
    "name": "Search Results",
    "items": [{
        "id": "https://example.edu/catalog/record/01234?query=IMS%20AND%20%28Caliper%20OR%20Analytics%29",
        "type": "Document",
        "name": "IMS Global Caliper Analytics Specification, version 1.1",
        "mediaType": "application/pdf"
      },
      {
        "id": "https://example.edu/catalog/record/09876?query=IMS%20AND%20%28Caliper%20OR%20Analytics%29",
        "type": "Document",
        "name": "IMS Global Caliper Analytics Implementation Guide, version 1.1",
        "mediaType": "application/pdf"
      },
      {
        "id": "https://example.edu/catalog/record/05432?query=IMS%20AND%20%28Caliper%20OR%20Analytics%29",
        "type": "VideoObject",
        "name": "Introduction to IMS Caliper",
        "mediaType": "video/ogg",
        "duration": "PT1H12M27S"
      }
    ],
    "dateCreated": "2018-08-01T06:00:00.000Z"
  }

  . . .

}
```

**Comment by arwhyte** [comment]
Below is a list of questions mentioned on the Tuesday, 5 June 2018 call, that require resolution:

1.  Should we list the optional `SearchEvent` properties that are not subject to value constraints?  No description, just a list (similar to what the Web Payments folks do; [example](https://web-payments.org/specs/source/vocabs/commerce.html#Account)).   I'm inclined to suggest we do so.   We'd do the same for `Query`. 

> ## Properties
>`id`, `type`, `actor`, `action`, `object`, `eventTime`, `target`, `generated`, `edApp`, `referrer`, `group`, `membership`, `session`, `federatedSession`, `extensions`.  Property value type constraints imposed by this profile are described below:

> \[Table\ follows] 

2.  Should value type of `SearchEvent.object` remain an `Entity` or be changed to `DigitalResource` or changed to `DigitalResource` | `SoftwareApplication`?

3.  Currently `Query` is a subtype of `Entity`.  Should it remain so or become a subtype of `DigitalResource`?

3. Should `Query.author` be changed to `Query.creator`?  Note that `DigitalResource` defines a `creators`\[<`Person`>\] property. 

4. Should `Query` be converted to a selector.  I need a concrete example of how this would be done.

5. Should we add a `SearchEvent.generated`<`DigitalResourceCollection`> as is described above?   If we do, should we subtype `DigitalResourceCollection` as `SearchResults`?

**Comment by ViktorHaag** [comment]
I'd rather have a cross reference to Event in the core spec, and say "all properties in core Event pertain; we document only constraints, additional fields, special cases" etc. People know how to use browser tabs, we don't need to add noise. (Assumes you _can_ cross-ref link to Event in the core spec, and if we can't I'd argue we _need to be able to do this_.)

**Comment by danielcgreen** [comment]
1. I like the list, but also recognize that this may introduce additional overhead in maintaining the document set if we ever add to the base class.

2. Prescribing the appropriate searchable objects feels presumptive.  I like the idea of leaving this open-ended.

3. Agree with @ViktorHaag 's POV.

4. I could be missing something, but the query entity itself seems overly verbose.  For me, the essense of the entity is the query string.  The search event seems to be the place for associating the actor and object with the search term.  Including these additional attributes in the query entity greatly increases the instances that will be persisted for receivers and I'm having trouble seeing the value in having unique query objects for each search that occurs between an actor and an object.  Rather, it seems we might consider containing only the id, type and searchString properties.

5. I don't have a strong opinion here, although a selector didn't feel obvious.

6. I'm not a huge fan of persisting search results.  While these would add context to the learner's subsequent navigation, I don't see it adding much value.

File: profiles/caliper-profile-search-v1p1.html [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+              be expressed either as an object or as a string corresponding to the actor's IRI.</td>
+            <td>Required</td>
+          </tr>
+          <tr>
+            <td>action</td>
+            <td><a href=
+               "https://www.imsglobal.org/sites/default/files/caliper/v1p1/caliper-spec-v1p1/caliper-spec-v1p1.html#termDef"
+            >Term</a></td>
+            <td>The action or predicate that binds the <code>actor</code> or subject to the
+                <code>object</code>. The value range is limited to the 'Searched' action only
```
Comment (danielcgreen) [inline profiles/caliper-profile-search-v1p1.html]: Should we include normative language here?  Perhaps "The value MUST be the "Searched" action."

Comment (arwhyte) [reply profiles/caliper-profile-search-v1p1.html]: @danielcgreen -- does it not already say that?

>The action or predicate that binds the <code>actor</code> or subject to the `object`. The value range is limited to the 'Searched' action only
              .

**Comment by arwhyte** [comment]
@ViktorHaag, @danielcgreen -- thanks for the comments.  I'll use Viktor's numbering in reply.

A1. Happy to withdraw the suggestion so no action on this one.

A2. Per @danielcgreen I've left the object as `Entity` unchanged.  @ViktorHaag how should one describe searching the UMich library site at (https://www.lib.umich.edu/) given our current vocabulary?   `SoftwareApplication` (i.e., some catalog) seems the more appropriate term to use as the `object` than `DigitalResource` given the former's definition: 

> A Caliper SoftwareApplication represents a computer program, application, module, platform or system.

A3.  Agreed.  `Query` should remain sub type of `Entity`.  BTW, I like @ViktorHaag's definition of digital resources: we should work the bolded phrasing into the spec:

> For me, "digital resources" **are persistent digital media that can present a locus of content that learners can consume while learning**: I don't think "query" falls into that category.

A4.  I changed `author` to `creator` to remain consistent with the current vocabulary.

A5.  It resembles a selector in a certain sense, yes.

> "like a non-contiguous selector applied to a digital resource, right?"

A6.  I've yet to add an optional generated `SearchResult`.  Want to talk to @danielcgreen about it on Wednesday.

**Comment by ViktorHaag** [comment]
A2 -- I don't think I have a problem referring to a Library as a digital resource, but it's uncomfortable that the Library's API entry point (i.e. the URL used to receive the search query) could end up as the same URI used to "identify" the library. Pragmatically, it's feasible to just treat this as a Software Application (although, still, the purist in me kind of rejects the idea that an Application is "searchable").

It seems to me the switching cost is lower if we start narrow and then expand later, rather than the other way around -- if we say "it's a DigitalResource or DigitalResourceCollection only" now, then we can add other things later (like SoftwareApplication); if we admit for things other than digital resources now, then we can't really take them off the table later.

**Comment by ViktorHaag** [comment]
Thinking again about Search Results. I think we need to draw a hard line between Digital Resources and the Search Results, both as a collection, and the items in the collection of search results.

It strikes me as highly likely that a search result _about_ a digital resource is _not_ the digital resource itself, but it might have a pointer to the digital resource. That is, the search result that contains meta-data tags, and an LTI link to the an item, isn't the item itself.

It's possible that a search result would send back the entire found object(s) themselves, but I highly doubt we want to start there -- we want to think of searches sending back _results_ which are things in and of themselves that tag and point to DR(C)s.

Once we start with that premise then we can:

- SearchResultCollection (caliper entity), which may or may not have an ID, and which contains a list of SearchResult things
- SearchResult (caliper entity), which I propose (as a start) has a defined sub-class that is an LRSearchResult (like we have Session-->LTISession)
- that LRSearchResult thing can have a close mapping to the actual LR Search terms so that we can crosswalk between Caliper and LR Search on the result payload level

We can also think about a SearchQuery caliper thing, with an LRSearchQuery sub-class, that does the cross walking to LRSearch's querying.

**Comment by arwhyte** [comment]
Merge once ontology is updated and fixtures are created.

**Comment by arwhyte** [comment]
caliper-contexts and caliper-ontology updated.  caliper-common-fixtures next.

**Comment by arwhyte** [comment]
caliper-common-fixtures updated.

**Pull request merged**